crates/engine/src/operation/solve/solver/shapes/mod.rs

```diff
@@ -22,352 +22,438 @@ use super::Solver;
+            );
+            included.toggle_range(..included.len());
+            included
+        };
 
-        let id = self.create_concrete_shape(exemplar, None, &data_fields, &typename_fields);
-        self.data_fields_buffer_pool.push(data_fields);
-        self.typename_fields_buffer_pool.push(typename_fields);
+        let shape_id = self.create_concrete_shape(
+            ObjectIdentifier::Anonymous,
+            None,
+            &typename_fields_sorted_by_response_key_str_then_position,
+            &data_fields_sorted_by_response_key_str_then_position,
+            included_typename_then_data_fields,
+        );
+        self.data_fields_buffer_pool
+            .push(data_fields_sorted_by_response_key_str_then_position);
+        self.typename_fields_buffer_pool
+            .push(typename_fields_sorted_by_response_key_str_then_position);
 
-        id
+        shape_id
     }
 
+    /// Create the expected shape with known expected fields, applying the GraphQL field collection
+    /// logic.
     fn create_concrete_shape(
         &mut self,
-        exemplar_object_id: ObjectDefinitionId,
-        maybe_response_object_set_id: Option<ResponseObjectSetDefinitionId>,
-        data_fields_sorted_by_response_key_then_position: &[DataField<'ctx>],
-        typename_fields_sorted_by_response_key_then_position: &[TypenameField<'ctx>],
+        identifier: ObjectIdentifier,
+        set_id: Option<ResponseObjectSetDefinitionId>,
+        typename_fields_sorted_by_response_key_str_then_position: &[TypenameField<'ctx>],
+        data_fields_sorted_by_response_key_str_then_position: &[DataField<'ctx>],
+        included_typename_then_data_fields: FixedBitSet,
     ) -> ConcreteShapeId {
-        let schema = self.schema;
-        tracing::trace!("Creating shape for exemplar {}", schema.walk(exemplar_object_id).name());
-
-        let typename_response_edges = typename_fields_sorted_by_response_key_then_position
-            .iter()
-            .filter(|field| match field.type_condition_id {
-                CompositeTypeId::Object(id) => id == exemplar_object_id,
-                CompositeTypeId::Interface(id) => {
-                    schema[id].possible_type_ids.binary_search(&exemplar_object_id).is_ok()
-                }
-                CompositeTypeId::Union(id) => schema[id].possible_type_ids.binary_search(&exemplar_object_id).is_ok(),
-            })
-            .dedup_by(|a, b| a.key.response_key == b.key.response_key)
-            .map(|field| field.key)
-            .collect();
-
-        let mut fields_buffer = self.data_fields_buffer_pool.pop();
         let mut field_shapes_buffer = self.field_shapes_buffer_pool.pop();
+        let mut distinct_typename_response_keys: Vec<PositionedResponseKey> = Vec::new();
+        let mut included = included_typename_then_data_fields.into_ones();
+
+        let mut all_expected_keys_equal_response_keys = true;
+        while let Some(i) = included.next() {
```

Comment:
It's an index, so at best I can do `index` or `ix` if you really think that's better. Otherwise I can add a comment if it's not clear what's going on for you